File: go/internal/memory/journal.go
// Append-only audit journal for the memory store.
//
// One JSON object per line, gzipped on rotation. Writes are best-effort:
// a failure to journal never breaks the underlying memory operation. Use
// Journal.Read(...) to stream historical entries with filtering and
// MemoryStore.Undo(entry) to reverse a single operation.
package memory

import (
	"bufio"
	"compress/gzip"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"math"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"
)

// JournalOp tags the kind of mutation that produced an entry. Free-form
// string so callers (reflect/decay/import) can introduce new ops.
type JournalOp = string

// JournalEntry is one line in the journal.
type JournalEntry struct {
	TS     float64                `json:"ts"`
	Op     JournalOp              `json:"op"`
	Actor  string                 `json:"actor"`
	ID     string                 `json:"id"`
	Before map[string]any         `json:"before,omitempty"`
	After  map[string]any         `json:"after,omitempty"`
	Meta   map[string]any         `json:"meta"`
}

// Summary returns a short human-readable description of the entry.
func (e JournalEntry) Summary() string {
	short := func(s string) string {
		if len(s) > 8 {
			return s[:8]
		}
		return s
	}
	clip := func(s string) string {
		if len(s) > 60 {
			return s[:60]
		}
		return s
	}
	switch e.Op {
	case "remember":
		if e.After != nil {
			if t, _ := e.After["text"].(string); t != "" {
				return fmt.Sprintf("remember %s «%s»", short(e.ID), clip(t))
			}
		}
	case "forget":
		if e.Before != nil {
			if t, _ := e.Before["text"].(string); t != "" {
				return fmt.Sprintf("forget %s «%s»", short(e.ID), clip(t))
			}
		}
	case "supersede":
		if e.Meta != nil {
			if nid, _ := e.Meta["new_id"].(string); nid != "" {
				return fmt.Sprintf("supersede %s → %s", short(e.ID), short(nid))
			}
		}
	case "link", "unlink":
		dst, _ := e.Meta["dst_id"].(string)
		rel, _ := e.Meta["rel"].(string)
		return fmt.Sprintf("%s %s → %s (%s)", e.Op, short(e.ID), short(dst), rel)
	}
	return fmt.Sprintf("%s %s", e.Op, short(e.ID))
}

func (e JournalEntry) toLine() ([]byte, error) {
	b, err := json.Marshal(e)
	if err != nil {
		return nil, err
	}
	return append(b, '\n'), nil
}

// Journal is an append-only JSONL log with size-based gzip rotation.
//
// Writes are atomic at the line level on POSIX (O_APPEND linearises across
// processes for writes ≤ PIPE_BUF). Reads tolerate truncated last lines (a
// crash mid-write) by skipping un-parseable JSON.
type Journal struct {
	Path             string
	RotateMB         int
	KeepDays         int
	writesSinceCheck int
}

const rotateCheckEvery = 256

// NewJournal constructs a Journal. The on-disk file is created lazily.
func NewJournal(path string, rotateMB, keepDays int) *Journal {
	if rotateMB <= 0 {
		rotateMB = 64
	}
	if keepDays <= 0 {
		keepDays = 90
	}
	return &Journal{Path: path, RotateMB: rotateMB, KeepDays: keepDays}
}

// Append writes one entry. All errors are logged, never returned, so a
// journal failure cannot break the underlying memory op.
func (j *Journal) Append(e JournalEntry) {
	if j == nil || j.Path == "" {
		return
	}
	if err := os.MkdirAll(filepath.Dir(j.Path), 0o755); err != nil {
		log.Printf("ai-houkai journal mkdir: %v", err)
		return
	}
	line, err := e.toLine()
	if err != nil {
		log.Printf("ai-houkai journal marshal: %v", err)
		return
	}
	f, err := os.OpenFile(j.Path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o644)
	if err != nil {
		log.Printf("ai-houkai journal open: %v", err)
		return
	}
	if _, err := f.Write(line); err != nil {
		log.Printf("ai-houkai journal write: %v", err)
	}
	_ = f.Close()

	j.writesSinceCheck++
	if j.writesSinceCheck >= rotateCheckEvery {
		j.writesSinceCheck = 0
		if err := j.maybeRotate(); err != nil {
			log.Printf("ai-houkai journal rotate: %v", err)
		}
	}
}

func (j *Journal) maybeRotate() error {
	if j.RotateMB > 0 {
		st, err := os.Stat(j.Path)
		if err == nil && st.Size() >= int64(j.RotateMB)*1024*1024 {
			if err := j.rotate(); err != nil {
				return err
			}
		}
	}
	if j.KeepDays > 0 {
		j.pruneArchives()
	}
	return nil
}

func (j *Journal) rotate() error {
	stamp := time.Now().UTC().Format("20060102T150405")
	stem := strings.TrimSuffix(filepath.Base(j.Path), filepath.Ext(j.Path))
	dir := filepath.Dir(j.Path)
	archive := filepath.Join(dir, fmt.Sprintf("%s-%s.log.gz", stem, stamp))
	if _, err := os.Stat(archive); err == nil {
		archive = filepath.Join(dir, fmt.Sprintf("%s-%s-%d.log.gz", stem, stamp, os.Getpid()))
	}
	src, err := os.Open(j.Path)
	if err != nil {
		return err
	}
	defer src.Close()
	dst, err := os.Create(archive)
	if err != nil {
		return err
	}
	gw := gzip.NewWriter(dst)
	if _, err := io.Copy(gw, src); err != nil {
		_ = gw.Close()
		_ = dst.Close()
		return err
	}
	if err := gw.Close(); err != nil {
		_ = dst.Close()
		return err
	}
	if err := dst.Close(); err != nil {
		return err
	}
	// Truncate in place so any concurrent writer's fd remains valid.
	return os.Truncate(j.Path, 0)
}

func (j *Journal) pruneArchives() {
	stem := strings.TrimSuffix(filepath.Base(j.Path), filepath.Ext(j.Path))
	pattern := filepath.Join(filepath.Dir(j.Path), stem+"-*.log.gz")
	matches, _ := filepath.Glob(pattern)
	cutoff := time.Now().Add(-time.Duration(j.KeepDays) * 24 * time.Hour)
	for _, p := range matches {
		st, err := os.Stat(p)
		if err == nil && st.ModTime().Before(cutoff) {
			_ = os.Remove(p)
		}
	}
}

// ReadOpts filters Journal.Read output.
type ReadOpts struct {
	Since           float64
	Until           float64
	Op              string
	Actor           string
	MemoryID        string
	Limit           int
	IncludeArchives bool
}

// Read streams matching entries oldest first.
func (j *Journal) Read(opts ReadOpts) ([]JournalEntry, error) {
	var files []string
	if opts.IncludeArchives {
		stem := strings.TrimSuffix(filepath.Base(j.Path), filepath.Ext(j.Path))
		pattern := filepath.Join(filepath.Dir(j.Path), stem+"-*.log.gz")
		archives, _ := filepath.Glob(pattern)
		sort.Strings(archives)
		files = append(files, archives...)
	}
	if _, err := os.Stat(j.Path); err == nil {
		files = append(files, j.Path)
	}

	var out []JournalEntry
	for _, fp := range files {
		f, err := os.Open(fp)
		if err != nil {
			continue
		}
		var r io.Reader = f
		if strings.HasSuffix(fp, ".gz") {
			gr, err := gzip.NewReader(f)
			if err != nil {
				_ = f.Close()
				continue
			}
			defer gr.Close()
			r = gr
		}
		sc := bufio.NewScanner(r)
		sc.Buffer(make([]byte, 0, 64*1024), 4*1024*1024)
		for sc.Scan() {
			line := strings.TrimSpace(sc.Text())
			if line == "" {
				continue
			}
			var e JournalEntry
			if err := json.Unmarshal([]byte(line), &e); err != nil {
				continue
			}
			if opts.Since != 0 && e.TS < opts.Since {
				continue
			}
			if opts.Until != 0 && e.TS > opts.Until {
				continue
			}
			if opts.Op != "" && e.Op != opts.Op {
				continue
			}
			if opts.Actor != "" && e.Actor != opts.Actor {
				continue
			}
			if opts.MemoryID != "" && e.ID != opts.MemoryID {
				continue
			}
			out = append(out, e)
			if opts.Limit > 0 && len(out) >= opts.Limit {
				_ = f.Close()
				return out, nil
			}
		}
		_ = f.Close()
	}
	return out, nil
}

// FindByTS returns the entry whose timestamp matches ts within tol (seconds).
func (j *Journal) FindByTS(ts, tol float64) (*JournalEntry, error) {
	if tol <= 0 {
		tol = 1e-3
	}
	entries, err := j.Read(ReadOpts{
		Since: ts - tol, Until: ts + tol, IncludeArchives: true,
	})
	if err != nil {
		return nil, err
	}
	for _, e := range entries {
		if math.Abs(e.TS-ts) <= tol {
			return &e, nil
		}
	}
	return nil, nil
}

File: ai_houkai/memory_system/journal.py
"""Append-only audit journal for the memory store.

One JSON object per line, gzipped on rotation. Writes are best-effort:
a failure to journal never breaks the underlying memory operation. Use
``Journal.read(...)`` to stream historical entries with filtering and
``MemoryStore.undo(entry)`` to reverse a single operation.
"""

from __future__ import annotations

import gzip
import json
import logging
import os
import time
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Iterator

log = logging.getLogger(__name__)


JournalOp = str  # "remember"|"forget"|"edit"|"supersede"|"restore"|"link"
                 # |"unlink"|"reflect"|"decay"|"import"|"export"|"undo"


@dataclass(frozen=True)
class JournalEntry:
    ts:     float
    op:     JournalOp
    actor:  str
    id:     str
    before: dict[str, Any] | None
    after:  dict[str, Any] | None
    meta:   dict[str, Any]

    def to_line(self) -> str:
        rec = {
            "ts": self.ts, "op": self.op, "actor": self.actor, "id": self.id,
            "before": self.before, "after": self.after, "meta": self.meta,
        }
        return json.dumps(rec, separators=(",", ":")) + "\n"

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> "JournalEntry":
        return cls(
            ts=float(d["ts"]),
            op=str(d["op"]),
            actor=str(d.get("actor", "lib")),
            id=str(d.get("id", "")),
            before=d.get("before"),
            after=d.get("after"),
            meta=d.get("meta") or {},
        )

    def summary(self) -> str:
        """Short, human-readable one-line description."""
        if self.op == "remember" and self.after:
            txt = (self.after.get("text") or "")[:60]
            return f"remember {self.id[:8]} «{txt}»"
        if self.op == "forget" and self.before:
            txt = (self.before.get("text") or "")[:60]
            return f"forget {self.id[:8]} «{txt}»"
        if self.op == "edit" and self.after:
            txt = (self.after.get("text") or "")[:60]
            return f"edit {self.id[:8]} «{txt}»"
        if self.op == "supersede":
            new_id = self.meta.get("new_id", "")
            return f"supersede {self.id[:8]} → {new_id[:8]}"
        if self.op in ("link", "unlink"):
            dst = self.meta.get("dst_id", "")
            rel = self.meta.get("rel", "")
            return f"{self.op} {self.id[:8]} → {dst[:8]} ({rel})"
        return f"{self.op} {self.id[:8]}"


class Journal:
    """Append-only JSONL journal with size-based gzip rotation.

    Writes are atomic at the line level on POSIX (O_APPEND linearises
    across processes for writes ≤ PIPE_BUF). Reads tolerate truncated
    last lines (a crash mid-write) by skipping un-parseable JSON.
    """

    _ROTATE_CHECK_EVERY = 256   # how often to size-check on append

    def __init__(
        self,
        path: str | Path,
        *,
        rotate_mb: int = 64,
        keep_days: int = 90,
    ) -> None:
        self.path = Path(path)
        self.rotate_mb = rotate_mb
        self.keep_days = keep_days
        self._writes_since_check = 0

    def append(self, entry: JournalEntry) -> None:
        """Append one entry. Failures are logged, not raised."""
        try:
            self.path.parent.mkdir(parents=True, exist_ok=True)
            with open(self.path, "a", encoding="utf-8") as f:
                f.write(entry.to_line())
        except OSError as e:
            log.warning("journal append failed: %s", e)
            return

        self._writes_since_check += 1
        if self._writes_since_check >= self._ROTATE_CHECK_EVERY:
            self._writes_since_check = 0
            try:
                self._maybe_rotate()
            except OSError as e:
                log.warning("journal rotate failed: %s", e)

    def _maybe_rotate(self) -> None:
        if self.rotate_mb > 0 and self.path.exists():
            size_mb = self.path.stat().st_size / (1024 * 1024)
            if size_mb >= self.rotate_mb:
                self._rotate()
        if self.keep_days > 0:
            self._prune_archives()

    def _rotate(self) -> None:
        stamp = time.strftime("%Y%m%dT%H%M%S")
        base = f"{self.path.stem}-{stamp}"
        rotated = self.path.with_name(f"{base}.log")
        archive = self.path.with_name(f"{base}.log.gz")
        if rotated.exists() or archive.exists():
            base = f"{base}-{os.getpid()}"
            rotated = self.path.with_name(f"{base}.log")
            archive = self.path.with_name(f"{base}.log.gz")
        # Rename first (atomic), then compress the renamed file. Appends that
        # land during/after the rename open the path fresh and so go to a new
        # active file — the old compress-then-truncate order silently dropped
        # any entry appended between the copy and the truncate. A crash
        # mid-compression leaves the plain rotated file behind; read() still
        # includes it, so no entries are lost either way.
        os.rename(self.path, rotated)
        with open(rotated, "rb") as src, gzip.open(archive, "wb") as dst:
            while True:
                chunk = src.read(1024 * 1024)
                if not chunk:
                    break
                dst.write(chunk)
        rotated.unlink()

    def _prune_archives(self) -> None:
        cutoff = time.time() - self.keep_days * 86_400
        # `-*.log` covers plain rotated files orphaned by a crash mid-compression.
        for pattern in (f"{self.path.stem}-*.log.gz", f"{self.path.stem}-*.log"):
            for p in self.path.parent.glob(pattern):
                try:
                    if p.stat().st_mtime < cutoff:
                        p.unlink()
                except OSError:
                    continue

    def read(
        self,
        *,
        since:     float | None = None,
        until:     float | None = None,
        op:        str | None = None,
        actor:     str | None = None,
        memory_id: str | None = None,
        limit:     int | None = None,
        include_archives: bool = False,
    ) -> Iterator[JournalEntry]:
        """Stream entries matching the filters, oldest first.

        Truncated / corrupted lines are silently skipped — the journal's
        purpose is best-effort forensics, not strict consistency.
        """
        files: list[Path] = []
        if include_archives:
            # Plain `-*.log` files are rotations orphaned by a crash before
            # compression finished — their entries are still valid history.
            files.extend(sorted(
                list(self.path.parent.glob(f"{self.path.stem}-*.log.gz"))
                + list(self.path.parent.glob(f"{self.path.stem}-*.log"))
            ))
        if self.path.exists():
            files.append(self.path)

        n_yielded = 0
        for fp in files:
            opener = gzip.open if fp.suffix == ".gz" else open
            try:
                handle = opener(fp, "rt", encoding="utf-8")
            except OSError:
                continue
            with handle as f:
                for line in f:
                    line = line.strip()
                    if not line:
                        continue
                    try:
                        entry = JournalEntry.from_dict(json.loads(line))
                    except (json.JSONDecodeError, KeyError, ValueError):
                        continue
                    if since is not None and entry.ts < since:
                        continue
                    if until is not None and entry.ts > until:
                        continue
                    if op is not None and entry.op != op:
                        continue
                    if actor is not None and entry.actor != actor:
                        continue
                    if memory_id is not None and entry.id != memory_id:
                        continue
                    yield entry
                    n_yielded += 1
                    if limit is not None and n_yielded >= limit:
                        return

    def find_by_ts(self, ts: float, *, tol: float = 1e-3) -> JournalEntry | None:
        """Locate the entry with timestamp matching *ts* (within *tol* seconds)."""
        for entry in self.read(since=ts - tol, until=ts + tol, include_archives=True):
            if abs(entry.ts - ts) <= tol:
                return entry
        return None
